readme.md
TuyaMQTT
==================

Listens on MQTT topic and routes requests to Tuya devices, based on a one to one topic translation. 

Quick Example
-----------
```
tuya/3.3/34280100600194d17c96/e7e9339aa82abe61/192.168.1.50/1/state
tuya/3.3/34280100600194d17c96/e7e9339aa82abe61/192.168.1.50/1/command
tuya/3.3/34280100600194d17c96/e7e9339aa82abe61/192.168.1.50/1/attributes
tuya/3.3/34280100600194d17c96/e7e9339aa82abe61/192.168.1.50/availability
tuya/3.3/34280100600194d17c96/e7e9339aa82abe61/192.168.1.50/attributes
```

Installation 
-----------
```
git clone https://github.com/TradeFace/tuyamqtt.git
cd tuyamqtt
make
make install
```

Installation Docker
-----------
```
git clone https://github.com/TradeFace/tuyamqtt.git
cd tuyamqtt
make docker
```

Running Docker
------------
```
docker run -it --rm --name my-app tuyamqtt
```

Running Docker Compose
-------------
```
version: '3'
services:
  homeassistant:
    ports: 
      - "8123:8123"
    ...
    restart: always
    network_mode: host
  mosquitto:
    image: eclipse-mosquitto
    ...
    restart: always
    network_mode: host
  tuyamqtt:
    image: "tuyamqtt:latest"
    hostname: tuyamqtt 
    container_name: tuyamqtt   
    working_dir: /usr/src/app    
    volumes:
      - ./config:/usr/src/app/config    
    command: "python -u main.py"
    restart: always
    network_mode: host
```

MQTT Topics
===========
```
tuya/<protocol>/<device-id>/<device-localkey>/<device-ip>/<dps>/state
tuya/<protocol>/<device-id>/<device-localkey>/<device-ip>/<dps>/command
tuya/<protocol>/<device-id>/<device-localkey>/<device-ip>/<dps>/attributes
tuya/<protocol>/<device-id>/<device-localkey>/<device-ip>/attributes
tuya/<protocol>/<device-id>/<device-localkey>/<device-ip>/availability
```

MQTT state
--------------
State payload values can be set in tuyamqtt.conf. Updates on command, or on state change on device (5 second polling)

- Default values: `ON`, `OFF`

MQTT command
--------------
Valid values are `true`, `ON`, `1`, `false`, `OFF`, `0`

MQTT dps attributes
--------------
Returns dps state, via state and timestamp

MQTT attributes
--------------
Returns all device attributes, via state and timestamp. Updates on command, or on state change on device (5 second polling)

MQTT availability
--------------
Availability payload values can be set in tuyamqtt.conf. Updates every 15 seconds. 

- Default values: `online`, `offline`.

Via state
-----------
The via state can be used in your home automation system. i.e. a state change with via `tuya` could be a strong indication of presence 
Returns `mqtt` when a state change comes in via a mqtt-request 
Returns `tuya` when a state change is detected in status polling 

tuyamqtt.conf
==============
```
[General]
entity_file: ./config/entities.json
topic: tuya
payload_on: ON
payload_off: OFF
availability_online: online
availability_offline: offline

[MQTT]
user: myusername
pass: mypassword
host: 192.168.1.14
port: 1883
```


Home Assistant Example
=============
```
switch:
  - platform: mqtt
    name: "Living Socket Left 1"
    command_topic: "tuya/3.3/34280100500194d17c95/e7e9339dd82abe61/192.168.1.50/1/command"  
    state_topic: "tuya/3.3/34280100500194d17c95/e7e9339dd82abe61/192.168.1.50/1/state"
    json_attributes_topic: "tuya/3.3/34280100500194d17c95/e7e9339dd82abe61/192.168.1.50/1/attributes"
    availability_topic: "tuya/3.3/34280100500194d17c95/e7e9339dd82abe61/192.168.1.50/availability"
  - platform: mqtt
    name: "Living Socket Left 2"
    command_topic: "tuya/3.3/347550642cf432a2cbaf/314c6b7f3f44f979/192.168.1.51/1/command"
    state_topic: "tuya/3.3/347550642cf432a2cbaf/314c6b7f3f44f979/192.168.1.51/1/state"
    json_attributes_topic: "tuya/3.3/347550642cf432a2cbaf/314c6b7f3f44f979/192.168.1.51/attributes"
    availability_topic: "tuya/3.3/347550642cf432a2cbaf/314c6b7f3f44f979/192.168.1.51/availability"
  - platform: mqtt
    name: "Living Socket Right 1"
    command_topic: "tuya/3.3/347550712cf432a2b04a/f9b80165db5e06bd/192.168.1.54/1/command" 
    state_topic: "tuya/3.3/347550712cf432a2b04a/f9b80165db5e06bd/192.168.1.54/1/state"
    json_attributes_topic: "tuya/3.3/347550712cf432a2b04a/f9b80165db5e06bd/192.168.1.54/attributes"
    availability_topic: "tuya/3.3/347550712cf432a2b04a/f9b80165db5e06bd/192.168.1.54/availability"
```
- Note: availability only works for devices known by TuyaMQTT.

Testing with Mosquitto
========================
Use mosquitto_pub to make TuyaMQTT aware of the device. TuyaMQTT needs a topic ending with `command` to discover a device

```mosquitto_pub -t tuya/3.1/34280100600194d17c96/e7e9339aa82abe61/192.168.1.50/1/command -m ON``` 

And run mosquitto_sub to listen to updates (in another terminal). 

```mosquitto_sub -t tuya/3.1/34280100600194d17c96/e7e9339aa82abe61/192.168.1.50/#```

It might take a few seconds to get a state update. 

Note: that TuyaMQTT only pushes status message when there was a change. The availability topic should be published every 15 sec. 

Debugging
================
```
python3 -u main.py -v
python3 -u main.py -vv
python3 -u main.py -vvv
```

TODO
===================
- check config values
- device config via topic is rather crude
  - simple interface?
- implement https://www.home-assistant.io/docs/mqtt/discovery/


Changelog
==================
- replaced entities.json with sqlite db
- only publish onchange
- added via mqtt/tuya
- thread per device
- pytuya replaced by https://github.com/TradeFace/tuya

Acknowledgements
=================
- https://github.com/SDNick484 for testing protocol 3.1 reimplementation

License
====================
Free to use, copy, modify, whatever you like.



web/changelog.md
# ChangeLog
-----------

## Changelog v0.1.0 - Initial setup
--------------

### general 
- dev: setup vue cli 3
- dev: added reststate
- dev: route to pages 


web.py
# from flask import Flask, jsonify, request, send_from_directory 
# from flask_restful import Resource, Api 
# from database import *

# app = Flask(__name__) 
# api = Api(app) 

# class Entities(Resource):

#     def get(self):

#         return jsonify(get_entities())

#     def post(self): 
          
#         data = request.get_json()     # status code 
#         return jsonify({'data': data}),

# class Settings(Resource):

#     def get(self):

#         return jsonify(get_settings())

# @app.route('/js/<path:path>')
# def send_js(path):
#     return send_from_directory('web/public/js', path)

# @app.route('/')
# def send_index():
#     return send_from_directory('web/public/html/','index.html')

# @app.route('/html/<path:path>')
# def send_html(path):
#     return send_from_directory('web/html', path)

# api.add_resource(Entities, '/api/entities') 
# api.add_resource(Settings, '/api/settings')

# app.run(debug = True) 

from flask import Flask, jsonify, request, send_from_directory 
from flask_cors import CORS, cross_origin
from flask_rest_jsonapi import Api, ResourceDetail, ResourceList, ResourceRelationship
from flask_sqlalchemy import SQLAlchemy
from marshmallow_jsonapi.flask import Schema, Relationship
from marshmallow_jsonapi import fields

# Create the Flask application and the Flask-SQLAlchemy object.
app = Flask(__name__)
app.config['DEBUG'] = True
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite://///home/niels/tools/tuyamqtt/config/tuyamqtt.db'
db = SQLAlchemy(app)

cors = CORS(app, resources={r"/api/*": {"origins": "*"}})

# Create model
class Settings(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    name = db.Column(db.Text)
    value = db.Column(db.Text)

class Entities(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    deviceid = db.Column(db.Text, unique=True)
    localkey = db.Column(db.Text)
    ip  = db.Column(db.Text)
    protocol  = db.Column(db.Text)
    topic  = db.Column(db.Text)
    attributes  = db.Column(db.Text)
    status_poll = db.Column(db.Float) 
    # status_command  = db.Column(db.Integer)
    # hass_discover = db.Column(db.Boolean)
    # name  = db.Column(db.Text)

# Create the database.
db.create_all()

# Create schema
class SettingsSchema(Schema):
    class Meta:
        type_ = 'setting'
        self_view = 'setting_detail'
        self_view_kwargs = {'id': '<id>'}
        self_view_many = 'setting_list'

    id = fields.Integer(as_string=True, dump_only=True)
    name = fields.Str()
    value = fields.Str()

class EntitiesSchema(Schema):
    class Meta:
        type_ = 'entity'
        self_view = 'entity_detail'
        self_view_kwargs = {'id': '<id>'}
        self_view_many = 'entity_list'

    id = fields.Integer(as_string=True, dump_only=True)    
    deviceid = fields.Str()
    localkey = fields.Str()
    ip = fields.Str()
    protocol = fields.Str()
    topic = fields.Str()
    attributes = fields.Str()
    status_poll = fields.Str()
    hass_discover = fields.Str()
# Create resource managers

class SettingsList(ResourceList):
    schema = SettingsSchema
    data_layer = {'session': db.session,
                  'model': Settings}

class SettingsDetail(ResourceDetail):
    schema = SettingsSchema
    data_layer = {'session': db.session,
                  'model': Settings}

class EntitiesList(ResourceList):
    schema = EntitiesSchema
    data_layer = {'session': db.session,
                  'model': Entities}

class EntitiesDetail(ResourceDetail):
    schema = EntitiesSchema
    data_layer = {'session': db.session,
                  'model': Entities}

# Create the API object
api = Api(app)
api.route(SettingsList, 'setting_list', '/api/settings')
api.route(SettingsDetail, 'setting_detail', '/api/settings/<int:id>')
api.route(EntitiesList, 'entity_list', '/api/entities')
api.route(EntitiesDetail, 'entity_detail', '/api/entities/<int:id>')

@app.route('/')
def send_index():
    return send_from_directory('web/dist/','index.html')

@app.route('/css/<path:path>')
def send_css(path):
    return send_from_directory('web/dist/css', path)

@app.route('/js/<path:path>')
def send_js(path):
    return send_from_directory('web/dist/js', path)

@app.route('/img/<path:path>')
def send_img(path):
    return send_from_directory('web/dist/img', path)

# Start the flask loop
if __name__ == '__main__':
    app.run()

web/src/router.js
import Vue from 'vue'
import Router from 'vue-router'
import Home from './views/home.vue'
import SettingList from '@/views/settinglist.vue'
import SettingDetail from '@/views/settingdetail.vue'
import EntityList from '@/views/entitylist.vue'
import EntityDetail from '@/views/entitydetail.vue'

Vue.use(Router)

export default new Router({
  mode: 'history',
  base: process.env.BASE_URL,
  routes: [
    {
      path: '/',
      name: 'home',
      component: Home
    },
    {
      path: '/settings',
      name: 'settinglist',
      component: SettingList
    },
    {
      path: '/settings/:id',
      name: 'settingdetail',
      component: SettingDetail
    },
    {
      path: '/entities',
      name: 'entitylist',
      component: EntityList
    },
    {
      path: '/entities/:id',
      name: 'entitydetail',
      component: EntityDetail
    },
  ]
})


tuyamqtt/database.py
import sqlite3
import json

db = sqlite3.connect('./config/tuyamqtt.db', check_same_thread=False)
cursor = db.cursor()

def disconnect():
    db.close()

def setup():
    
    cursor.execute('''
        CREATE TABLE IF NOT EXISTS entities (id INTEGER PRIMARY KEY, deviceid TEXT unique,
                        localkey TEXT, ip TEXT, protocol TEXT, topic TEXT, attributes TEXT, status_poll FLOAT, hass_discover BOOL)
    ''')  
    cursor.execute('''
        CREATE TABLE IF NOT EXISTS settings (id INTEGER PRIMARY KEY, name TEXT unique, value TEXT)
    ''')  
    db.commit()

    #not used yet
    settings = [
        {'name': 'mqtt_host', 'value': '192.168.1.14'},
        {'name': 'mqtt_port', 'value': '1883'},
        {'name': 'discovery_topic', 'value': 'tuya'}
    ]
    insert_settings(settings)
    # get_settings()

def get_settings():

    cursor.execute('''SELECT * FROM settings''')
    all_rows = cursor.fetchall()
    print(all_rows)
    return all_rows

def insert_setting(setting:dict):
    try:
        cursor.execute('''INSERT INTO settings(name, value)
                        VALUES(:name, :value)''',
                        setting)
        db.commit()       
    except Exception as e:
        # print(e)
        db.rollback()
        return False    
    return True

def insert_settings(settings:list):

    if False in set(map(insert_setting, settings)):
        return False 
    return True

#quick and dirty
def get_entities():

    dictOfEntities = {} 
    cursor.execute('''SELECT * FROM entities''')
    all_rows = cursor.fetchall()
    for row in all_rows:
        
        entity = {
            'id': row[0],
            'deviceid': row[1],
            'localkey': row[2],            
            'ip': row[3],
            'protocol': row[4],
            'topic': row[5],
            'attributes': json.loads(row[6]),
            'status_poll': row[7],
            'hass_discover': row[8]
        }
        dictOfEntities[row[1]] = entity
    # print(dictOfEntities)
    return dictOfEntities

def attributes_to_json(entity:dict):

    dbentity = dict(entity)
    dbentity['attributes'] = json.dumps(dbentity['attributes'])
    return dbentity

def insert_entity(entity:dict):
    # print('insert_entity')
    try:
        cursor.execute('''INSERT INTO entities(deviceid, localkey, ip, protocol, topic, attributes, status_poll)
                        VALUES(:deviceid, :localkey, :ip, :protocol, :topic, :attributes, :status_poll)''',
                        attributes_to_json(entity))
        db.commit()
        entity['id'] = cursor.lastrowid
    except Exception as e:
        # print(e)
        db.rollback()
        return False
    
    return True
    #insert attributes
    # db.commit()

def update_entity(entity:dict):
    # print('update_entity',entity)
    # dbentity = attributes_to_json(entity)
    try:
        with db:
            db.execute('''UPDATE entities 
                    SET deviceid = ?, localkey = ?, ip = ?, protocol = ?, topic = ?, attributes = ?, status_poll = ?
                    WHERE id = ?''',
                    (entity['deviceid'], entity['localkey'], entity['ip'], entity['protocol'], entity['topic'], json.dumps(entity['attributes']), entity['status_poll'], entity['id'])
                    )
    except Exception as e:
        # print(e)
        return False
    return True


def upsert_entity(entity:dict):
    # print(entity)      
    if not insert_entity(entity):
        return update_entity(entity)


def upsert_entities(entities:dict):
 
    if False in set(map(upsert_entity, entities.values())):
        return False 
    return True
    

def delete_entity(entity:dict):

    cursor.execute('''DELETE FROM entities WHERE id = ? ''', (entity['id'],))
    #delete attributes
    db.commit()


tuyamqtt/__init__.py
import time
import paho.mqtt.client as mqtt
import json
import tuya
from os import path
from threading import Thread
import database as database

def connack_string(state):

    states = [
        'Connection successful',
        'Connection refused - incorrect protocol version',
        'Connection refused - invalid client identifier',
        'Connection refused - server unavailable',
        'Connection refused - bad username or password',
        'Connection refused - not authorised'
    ]
    return states[state]


def payload_bool(payload:str):

        str_payload = str(payload.decode("utf-8"))
        if str_payload == 'True' or str_payload == 'ON' or str_payload == '1':
            return True       
        elif str_payload == 'False' or str_payload == 'OFF' or str_payload == '0':
            return False    
        return payload


def bool_payload(config:dict, boolvalue:bool):

    if boolvalue:
        return config['General']['payload_on']
    return config['General']['payload_off']


def bool_availability(config:dict, boolvalue:bool):

    if boolvalue:
        return config['General']['availability_online']
    return config['General']['availability_offline']


class TuyaMQTTEntity(Thread):

    delay = 0.1

    def __init__(self, key, entity, parent):        
 
        Thread.__init__(self)
        self.key = key
        # self.mqtt_topic = key
        self.entity = entity
        self.mqtt_topic = entity['topic']
        # if entity['topic'] == '':
        self.mqtt_topic = "tuya/%s/%s/%s/%s"%(entity['protocol'],entity['deviceid'],entity['localkey'],entity['ip'])

        self.parent = parent
        self.config = self.parent.config   
        self.debuglevel =  self.parent.debuglevel
        self.mqtt_connected = False
        self.availability = False
        self.availability_changed = False


    def mqtt_connect(self): 

        try:
            self.mqtt_client = mqtt.Client()
            self.mqtt_client.enable_logger()
            self.mqtt_client.username_pw_set(self.config['MQTT']['user'], self.config['MQTT']['pass'])
            self.mqtt_client.connect(self.config['MQTT']['host'], int(self.config['MQTT']['port']), 60)
            self.mqtt_client.on_connect = self.on_connect
            self.mqtt_client.loop_start()   
            self.mqtt_client.on_message = self.on_message
        except Exception as ex:
            print('Failed to connect to MQTT Broker', ex)
            self.mqtt_connected = False


    def on_message(self, client, userdata, message):      

        if message.topic[-7:] != 'command':
            return   

        if self.debuglevel >= 1:
            print("topic",message.topic,"retained",message.retain,"message received",str(message.payload.decode("utf-8")))

        entityParts = message.topic.split("/")  
        dps_key = str(entityParts[5]) #will give problem with custom topics

        if dps_key not in self.entity['attributes']['dps']:
            self._set_dps(dps_key, None)
        if dps_key not in self.entity['attributes']['via']:
            self._set_via(dps_key, 'mqtt')
        self.set_status(dps_key, payload_bool(message.payload))

    def _set_dps(self, dps_key, dps_value:str):

        self.entity['attributes']['dps'][dps_key] = dps_value
        self.parent.set_entity_dps_item(self.key, dps_key, dps_value) 

    def _set_via(self, dps_key, via:str):

        self.entity['attributes']['via'][dps_key] = via
        self.parent.set_entity_via_item(self.key, dps_key, via) 

    def _set_availability(self, availability):

        if availability != self.availability:
            self.availability = availability
            self.availability_changed = True

    def _process_data(self, data:dict, via:str, force_mqtt:bool = False):

        changed = force_mqtt
        
        for dps_key, dps_value in data['dps'].items():
            
            if dps_key not in self.entity['attributes']['dps']:
                self._set_dps(dps_key, None)
            if self.debuglevel >= 3:
                print("_process_data ", dps_key, dps_value)
            if dps_key not in self.entity['attributes']['via']:
                self._set_via(dps_key, 'init')
            if dps_value != self.entity['attributes']['dps'][dps_key] or force_mqtt:
                changed = True
                self._set_dps(dps_key, dps_value) 
                if self.debuglevel >= 3:
                    print("->publish %s/%s/state" % (self.mqtt_topic, dps_key))                
                self.mqtt_client.publish("%s/%s/state" % (self.mqtt_topic, dps_key),  bool_payload(self.config, dps_value))  
                
                if via != self.entity['attributes']['via'][dps_key]:                        
                    self._set_via(dps_key, via)
                
                attr_item = {
                    'dps': self.entity['attributes']['dps'][dps_key], 
                    'via': self.entity['attributes']['via'][dps_key],
                    'time': time.time()
                }
                if self.debuglevel >= 3:
                    print("->publish %s/%s/attributes" % (self.mqtt_topic, dps_key)) 
                self.mqtt_client.publish("%s/%s/attributes" % (self.mqtt_topic, dps_key),  json.dumps(attr_item))
                # print("%s/%s/attributes" % (self.mqtt_topic, dps_key))
        
        if changed:
            attr = {
                'dps': self.entity['attributes']['dps'], 
                'via': self.entity['attributes']['via'],
                'time': time.time()
            } 
            if self.debuglevel >= 3:
                    print("->publish %s/attributes" % (self.mqtt_topic))
            self.mqtt_client.publish("%s/attributes" % (self.mqtt_topic),  json.dumps(attr))


    def status(self, via:str = 'tuya', force_mqtt:bool = False):
            
        try:
            data = tuya.status(self.entity)

            if not data:
                self._set_availability(False)
                return
       
            self._process_data(data, via, force_mqtt)
            self._set_availability(True)

        except Exception as ex:
            print(ex, 'status for', self.mqtt_topic)
            self._set_availability(False)


    def set_status(self, dps_item, payload):
        # print('set_status')
        try:  
            timer = time.time()
            data = tuya.set_status(self.entity, dps_item, payload)            
            # print(self.mqtt_topic,data, time.time()-timer)
            if data == None:
                self.status('mqtt', True)
                return

            self._process_data(data, 'mqtt', True)

        except Exception as ex:
            print(ex, 'set_status for', self.mqtt_topic)

    def hass_discovery(self):

        hass_topic = 'homeassistant/%s/%s'

        dps = 1
        payload = {
            "name": "name here",
            "cmd_t": "~command",
            "stat_t": "~state",
            "val_tpl":"{{value_json.POWER}}",
            "pl_off": self.config['General']['payload_off'],
            "pl_on": self.config['General']['payload_off'],
            "avty_t":"~availability",
            "pl_avail": self.config['General']['availability_online'],
            "pl_not_avail": self.config['General']['availability_offline'],
            "uniq_id": '%s_%s' % (self.entity['deviceid'], dps),
            "device":{
                "identifiers":[self.entity['deviceid']],
                "connections":[["mac", "D8:F1:5B:8C:60:4F"]]
            },
            "~": self.mqtt_topic
        }
        print(payload)

    def run(self):

        time_run_availability = 0
        time_run_status = 0
        # time_unset_reset = 0  
        # self.hass_discovery()

        while True:  

            if not self.mqtt_connected:
                self.mqtt_connect()
                time.sleep(1)         

            if time.time() > time_run_status:   
                if self.debuglevel >= 2:
                    print('->status poll '+self.entity['ip']) 
                self.status()                
                time_run_status = time.time()+self.entity['status_poll']  
                if self.debuglevel >= 2: 
                    print('<-status poll '+self.entity['ip'])             

            if time.time() > time_run_availability:               
                time_run_availability = time.time()+15   
                if self.debuglevel >= 3:
                    print("->publish %s/availability" % self.mqtt_topic)     
                self.mqtt_client.publish("%s/availability" % self.mqtt_topic, bool_availability(self.config, self.availability))         

            time.sleep(self.delay)            

   
    def on_connect(self, client, userdata, flags, rc):

        if self.debuglevel >= 1:
            print("MQTT Connection state: %s for %s" % (connack_string(rc), self.mqtt_topic))
        client.subscribe("%s/#" % self.mqtt_topic)
        self.mqtt_connected = True


class TuyaMQTT:


    delay = 0.1
    config = []
    dictOfEntities = {}


    def __init__(self, config):

        self.config = config

        self.entities_file = config['General']['entity_file']
        self.mqtt_topic = config['General']['topic']
        self.mqtt_connected = False

        self.debuglevel = 0
        if config['General']['debug'] == '-v':
            self.debuglevel = 1
        elif config['General']['debug'] == '-vv':
            self.debuglevel = 2
        elif config['General']['debug'] == '-vvv':
            self.debuglevel = 3

        self.database = database
        self.database.setup()  
        #debug 3.1
        # self.add_entity_dict('tuya/3.1/06200290b4e62d195a42/a52e118707bf530a/192.168.1.143', True)      


    def mqtt_connect(self): 

        try:
            self.mqtt_client = mqtt.Client()
            self.mqtt_client.enable_logger()
            self.mqtt_client.username_pw_set(self.config['MQTT']['user'], self.config['MQTT']['pass'])
            self.mqtt_client.connect(self.config['MQTT']['host'], int(self.config['MQTT']['port']), 60)
            self.mqtt_client.on_connect = self.on_connect
            self.mqtt_client.loop_start()   
            self.mqtt_client.on_message = self.on_message
        except Exception as ex:
            print('Failed to connect to MQTT Broker', ex)
            self.mqtt_connected = False
   

    def on_connect(self, client, userdata, flags, rc):

        if self.debuglevel >= 1:
            print("MQTT Connection state: %s for %s" % (connack_string(rc), self.mqtt_topic))
        client.subscribe("%s/#" % self.mqtt_topic)
        self.mqtt_connected = True


    def write_entity(self):

        # try:
        #     with open(self.entities_file, 'w') as fp:
        #         json.dump(self.dictOfEntities, fp, indent=4)
        # except:
        #     print("%s not writable" % self.entities_file)  
        self.database.upsert_entities(self.dictOfEntities)


    def read_entity(self):

        # if not path.isfile(self.entities_file):
        #     print('file not found')
        #     return

        # with open(self.entities_file) as fp:
        #     self.dictOfEntities = json.loads(fp.read())

        self.dictOfEntities = self.database.get_entities()


    def add_entity_dict(self, entityRaw, retain):
        
        entityParts = entityRaw.split("/")

        key = entityParts[2]

        if key in self.dictOfEntities:
            return False
        
        entity = {
            'protocol': entityParts[1],
            'deviceid': entityParts[2],
            'localkey': entityParts[3],
            'ip': entityParts[4],
            'topic': key,            
            'attributes': {
                'dps': {},
                'via': {}
            },
            'status_poll': 5.0,
            'hass_discover': False
        }

        self.dictOfEntities[key] = entity
        # self.write_entity()
        self.database.insert_entity(entity)
        return key

    def get_entity(self, key):

        return self.dictOfEntities[key]


    def set_entity_dps_item(self, key, dps, value):

        self.dictOfEntities[key]['attributes']['dps'][dps] = value   
        self.database.update_entity(self.dictOfEntities[key])


    def set_entity_via_item(self, key, dps, value):

        self.dictOfEntities[key]['attributes']['via'][dps] = value
        self.database.update_entity(self.dictOfEntities[key])
    

    def on_message(self, client, userdata, message):                   

        if message.topic[-7:] != 'command':
            return   
        
        key = self.add_entity_dict(message.topic, message.retain)

        if key:
            if self.debuglevel >= 1:
                print("message received",str(message.payload.decode("utf-8")),\
                "topic",message.topic,"retained",message.retain)    
            entity = self.get_entity(key)
            
            myThreadOb1 = TuyaMQTTEntity(key, entity, self)     
            myThreadOb1.setName(key)    
            myThreadOb1.start()


    def main_loop(self):

        """
        primary loop to send / receive from tuya devices
        """
        self.read_entity()
     
        tpool = []
        for key,entity in self.dictOfEntities.items():
            myThreadOb1 = TuyaMQTTEntity(key, entity, self)     
            myThreadOb1.setName(key)    
            myThreadOb1.start()
            tpool.append(myThreadOb1)
 
        time_run_save = 0
        
        while True: 

            if not self.mqtt_connected:
                self.mqtt_connect()
                time.sleep(2)
                continue                      
          
            if time.time() > time_run_save:
                self.write_entity()
                time_run_save = time.time()+300

            time.sleep(self.delay)           

